lib/qesapdeployment.pm

```diff
@@ -91,7 +101,122 @@ sub qesap_get_deployment_code {
+}
+
+
+=head3 qesap_upload_logs
+
+    qesap_upload_logs([failok=1])
+
+    Collect and upload logs present in @log_files.
+    failok - continue even in case upload fails
+
+=cut
+
+sub qesap_upload_logs {
+    my (%args) = @_;
+    my $failok = $args{failok} ? $args{failok} : 0;
+    record_info("Uploading logfiles", join("\n", @log_files));
+    for my $file (@log_files) {
+        upload_logs($file, failok => $failok);
+    }
+    # Remove already uploaded files from arrays
+    @log_files = ();
+}
+
+
+=head3 qesap_yaml_replace
+
+    Replaces yaml config file variables with parameters defined by OpenQA testode, yaml template or yaml schedule.
+    Openqa variables need to be added as a hash with key/value pair inside %run_args{openqa_variables}.
+    Example:
+        my %variables;
+        $variables{HANA_SAR} = get_required_var("HANA_SAR");
+        $variables{HANA_CLIENT_SAR} = get_required_var("HANA_CLIENT_SAR");
+        qesap_yaml_replace(openqa_variables=>\%variables);
+=cut
+
+sub qesap_yaml_replace {
+    my (%args) = @_;
+    my $variables = $args{openqa_variables};
+    my %variables = %$variables;
+    my %replaced_variables = ();
+
+    for my $variable (keys %variables) {
+        $replaced_variables{"%" . $variable . "%"} = %variables{$variable};
+    }
+    file_content_replace(QESAP_CONF_TRGT, %replaced_variables);
+
+    push(@log_files, QESAP_CONF_TRGT);
+    qesap_upload_logs();
 }
 
 
+=head3 qesap_execute
+
+    qesap_execute(cmd => $qesap_script_cmd [, verbose => 1] );
+
+    Execute qesap glue script commands. Check project documentation for available options:
+    https://github.com/SUSE/qe-sap-deployment
+=cut
+
+sub qesap_execute {
+    my (%args) = @_;
+    my $verbose = $args{verbose} ? "--verbose" : "";
+    my $qesap_cmd = DEPLOYMENT_DIR . "/scripts/qesap/qesap.py " . $verbose . " -c " . QESAP_CONF_TRGT . " -b " . DEPLOYMENT_DIR;
+    my $command = defined($args{cmd}) ? $args{cmd} : die("QESAP command to execute undefined");
+    my $exec_log = "/root/qesap_exec_" . $command . ".log.txt";
+
+    record_info('QESAP exec', 'Executing: \n' . $qesap_cmd . " " . $command);
+    push(@log_files, $exec_log);
+    assert_script_run($qesap_cmd . " " . $command . " > " . $exec_log, quiet => 1);
+    qesap_upload_logs();
+}
+
+
+=head3 qesap_prepare_env
+
+    qesap_prepare_env(variables=>{dict with variables});
+
+    Prepare terraform environment.
+    - creates file structures
+    - pulls git repository
+    - external config files
+    - installs pip requirements and OS packages
+    - generates config files with qesap script
+
+    For variables example see 'qesap_yaml_replace'
+=cut
+
+sub qesap_prepare_env {
+    my (%args) = @_;
+    my $variables = $args{openqa_variables};
+    my $pip_cmd = "pip install --no-color --no-cache-dir --progress-bar off ";
+    my $tfvars_template = get_var('QESAP_TFVARS_TEMPLATE');
+    my $curl = "curl -v -L ";
+
```

Comment:
maybe here you like to call qesap_create_folder_tree ?

Reply:
Hmm... that is a good point. But that might be only needed if not using git clone. 
Git clone does create the directory by itself so somewhere in qesap_get_deployment_code.